File: Data Structures & Algorithms/search-for-word/submission-1.rs
impl Solution {
    pub fn exist(mut board: Vec<Vec<char>>, word: String) -> bool {
        let rows = board.len();
        let cols = board[0].len();
        let words: Vec<char> = word.chars().collect();

        fn dfs(
            board: &mut Vec<Vec<char>>, 
            word: &[char],
            row: i32, 
            col: i32, 
            idx: usize,
        ) -> bool {
                if idx == word.len() {
                    return true;
                }

                if row < 0
                    || row >= board.len() as i32
                    || col < 0
                    || col >= board[0].len() as i32
                    || board[row as usize][col as usize] != word[idx] {
                    return false;
                }

                let ori = board[row as usize][col as usize];
                board[row as usize][col as usize] = '#';

                let found = dfs(board, word, row-1, col, idx+1)
                    || dfs(board, word, row+1, col, idx+1)
                    || dfs(board, word, row, col-1, idx+1)
                    || dfs(board, word, row, col+1, idx+1);

                board[row as usize][col as usize] = ori;

                found
        }

        for row in 0..rows {
            for col in 0..cols {
                if dfs(
                    &mut board, &words, row as i32, col as i32, 0
                ) {
                    return true;
                }
            }
        }
        false
    }
}

File: Data Structures & Algorithms/valid-sudoku/submission-1.go
func isValidSudoku(board [][]byte) bool {
    rows := make([]map[byte]bool, 9)
    cols := make([]map[byte]bool, 9)
    boxes := make([]map[byte]bool, 9)


    for i := 0; i < 9; i++ {
        rows[i] = make(map[byte]bool)
        cols[i] = make(map[byte]bool)
        boxes[i] = make(map[byte]bool)
    }

    for r := 0; r < 9; r++ {
        for c := 0; c < 9; c++ {
            val := board[r][c]

            if val == '.' {
                continue
            }

            box := (r/3)*3 + c/3

            if rows[r][val] || cols[c][val] || boxes[box][val] {
                return false
            }

            rows[r][val] = true
            cols[c][val] = true
            boxes[box][val] = true
        }
    }

    return true
}
